kyo-core/shared/src/main/scala/kyo/Meter.scala

```diff
@@ -225,16 +221,126 @@ object Meter:
 
                 def close(using Frame): Boolean < IO =
                     Kyo.foreach(meters)(_.close).map(_.exists(identity))
+
+                def closed(using Frame): Boolean < IO =
+                    Kyo.foreach(meters)(_.closed).map(_.exists(identity))
             end new
         }
 
-    private def offer[A](n: Int, chan: Channel[A], v: A)(using Frame): Unit < IO =
-        Loop.indexed { idx =>
-            if idx == n then Loop.done
+    abstract private class Base(permits: Int)(using initFrame: Frame, allow: AllowUnsafe) extends Meter:
+
+        // MinValue => closed
+        // >= 0     => # of permits
+        // < 0      => # of waiters
+        val state   = AtomicInt.Unsafe.init(permits)
+        val waiters = new MpmcUnboundedXaddArrayQueue[Promise.Unsafe[Closed, Unit]](8)
+        val closed  = Promise.Unsafe.init[Closed, Nothing]()
+
+        def dispatch[A, S](v: => A < S): A < (S & IO)
+
+        final def run[A, S](v: => A < S)(using Frame) =
+            IO.Unsafe {
+                @tailrec def loop(): A < (S & Async & Abort[Closed]) =
+                    val st = state.get()
+                    if st == Int.MinValue then
+                        // Meter is closed
+                        closed.safe.get
+                    else if state.cas(st, st - 1) then
+                        if st > 0 then
+                            // Permit available, dispatch immediately
+                            dispatch(v)
+                        else
+                            // No permit available, add to waiters queue
+                            val p = Promise.Unsafe.init[Closed, Unit]()
+                            waiters.add(p)
+                            p.safe.use(_ => dispatch(v))
+                    else
+                        // CAS failed, retry
+                        loop()
+                    end if
+                end loop
+                loop()
+            }
+        end run
+
+        final def tryRun[A, S](v: => A < S)(using Frame): Maybe[A] < (S & Async & Abort[Closed]) =
```

Comment:
I considered that, `offer` in `Channel` and `Queue` is a similar scenario. Aborting seems a safer behavior since it could highlight potential issues but I don't have much experience with tracking failures statically. @steinybot could you share your thoughts?